opentelemetry-api/src/opentelemetry/metrics/__init__.py

```diff
@@ -11,3 +11,513 @@
+
+        Args:
+            distributed_context: Explicit :class:`.DistributedContext` to use
+            instead of the current context. Context is used to add dimensions
+            for the resulting metric calculated out of the provided
+            measurements.
+            span_context: the :class:`.SpanContext` that identified the
+            :class:`.Span` for which the measurements are associated with.
+        """
+
+
+class Measurement:
+    """An empty interface that represents a single value.
+
+    This single value is recorded for the :class:`.Measure` that created
+    this measurement.
+    """
+
+
+class Measure:
+
+    def __init__(self,
+                 name: str,
+                 options: typing.Optional['MeasureOptions'] = None) -> None:
+        """Used to create raw :class:`.Measurement`s.
+
+        A contract between the API exposing the raw measurement and SDK
+        aggregating these values into the :class:`.Metric`. Measure is
+        constructed from the :class:`.Meter` class by providing a set of
+        :class:`.MeasureOptions`.
+        """
+        self.name = name
+        if options:
+            self.description = options.description
+            self.unit = options.unit
+            self.measure_type = options.measure_type
+
+    def create_double_measurement(self,
+                                  value: float) -> 'Measurement':
+        """Creates a measurement that contains double values.
+
+        Args:
+            value: The value of the measurement.
+
+        Returns:
+            A new :class:`.Measurement`
+        """
+
+    def create_long_measurement(self,
+                                value: int) -> 'Measurement':
+        """Creates a measurement that contains long values.
+
+        Args:
+            value: The value of the measurement.
+
+        Returns:
+            A new :class:`.Measurement`
+        """
+
+
+class Metric:
+    """Base class for various types of metrics.
+
+    Metric class that inherit from this class are specialized with the type of
+    time series that the metric holds. Metric is constructed from the
+    :class:`.Meter` class, by providing a set of :class:`.MetricOptions`.
+    """
+
+    def get_or_create_time_series(self,
+                                  label_values: typing.List['LabelValue']
+                                  ) -> 'object':
+        """Gets and returns a `TimeSeries`, a container for a cumulative value.
+
+        If the provided label values are not already associated with this
+        metric, a new timeseries is returned, otherwise it returns the existing
+        timeseries with the exact label values. The timeseries returned
+        contains logic and behaviour specific to the type of metric that
+        overrides this function.
+
+        Args:
+            label_values: A map of :class:`.LabelValue`s that will be
+            associated with the return timeseries.
+        """
+        raise NotImplementedError
```

Comment:
I don't have a strong opinion here. If it's worth including in the API I think it's fine to require implementers to override it.

FWIW I didn't raise this in the trace package because we wanted to avoid raising any unhandled exceptions. Both ABCs and `NotImplementedError` have this risk, but ABCs cause the error sooner.

Reply:
Yeah, I could generally see things in the API being absolutely required, even if we make backwards incompatible changes. I guess I would add an action item here to change this to an ABC for consistencies sake.